### PR TITLE
Clang format integration

### PR DESCRIPTION
**Pull request by cugone**
I loaded a new page and lost all my notes so I'm not going to re-type it all >.<

Suffice to say, clang-format is supported in VS2017 (v6.0) and VS2019 (v7.0). I added the files necessary to uphold the coding conventions documented in CONVENTIONS.md as close as possible. Deviations are commented and commented out for when using the tool in later VS versions.

This will be especially helpful with tabs and brace styles.

See VS blog: https://devblogs.microsoft.com/cppblog/clangformat-support-in-visual-studio-2017-15-7-preview-1/ for more info.

### CODE REVIEW

**Comment by Brett208** [comment]
I was reviewing this branch. The changes made to the file proj/vs2017/NAS2D.vcxproj.filters are lost in the noise. I suspect the original file had Windows specific line endings. Github is likely forcing those line endings over to Linux style. Maybe it would be better to try and separate the commit into one that only applies the line ending change and one that applies the meat of the change to the filters file.

I think Dan has had good success with does2unix to complete this in the past. https://waterlan.home.xs4all.nl/dos2unix.html

I am very guilty of making others clean up my line ending commits in the past. :|

**Comment by DanRStevens** [comment]
I checked into this briefly. The Visual Studio project and filter files were both using CRLF (Windows style) line endings. When this change was made, the filter file was converted to LF (Linux style) line endings.

I suspect it was done automatically by the Git configuration. You may need to adjust Git settings to make the commit without it getting auto converted. (If you undid the change with `dos2unix`, or rather `unix2dos`, it might be redone by Git unless settings are changed).

Some references:
https://help.github.com/en/articles/dealing-with-line-endings
https://stackoverflow.com/questions/170961/whats-the-best-crlf-carriage-return-line-feed-handling-strategy-with-git

There are some strong opinions on this issue, and no real consensus. I prefer the version control system not make silent automatic conversions on me, though I understand there can be some practical implications of this when working with Visual Studio. I believe Visual Studio can handle LF endings just fine for project and filter files, so you're free to convert if you want.

As Brett has said, it's strongly preferred to convert line endings separately in their own commit. This is to keep functional changes from getting lost in the noise of the EOL conversion.

**Comment by cugone** [comment]
90% of the time when I pull latest Visual Studio warns me "Line endings aren't consistent" and I just click through to change them back to windows CRLF (because why would a linux environment be using a VS config file?) and I guess I clicked No for once. :/

File: proj/vs2017/NAS2D.vcxproj.filters [hunk 1]
```diff
@@ -135,9 +135,6 @@
     <ClCompile Include="..\..\src\Xml\XmlNode.cpp">
       <Filter>Source Files\Xml</Filter>
     </ClCompile>
-    <ClCompile Include="..\..\src\Mixer\Mixer.cpp">
-      <Filter>Source Files\Mixer</Filter>
-    </ClCompile>
```
Comment (DanRStevens) [inline proj/vs2017/NAS2D.vcxproj.filters]: This is a surprising and unrelated change.

File: proj/vs2017/NAS2D.vcxproj.filters [hunk 2]
```diff
@@ -272,12 +269,9 @@
     <ClInclude Include="..\..\include\NAS2D\Xml\XmlMemoryBuffer.h">
       <Filter>Header Files\Xml</Filter>
     </ClInclude>
-    <ClInclude Include="..\..\include\NAS2D\BuildConfig.h">
-      <Filter>Header Files</Filter>
-    </ClInclude>
```
Comment (DanRStevens) [inline proj/vs2017/NAS2D.vcxproj.filters]: This is also surprising and unrelated.

**Comment by DanRStevens** [comment]
In general, please avoiding porting unrelated changes across branches, particularly when they are not required by the unrelated branch.

Perhaps more confusing, is the deleted code does not exist on master. I think what happened is the commit that modified the filters file added those lines, only we couldn't see that because of all the EOL noise.

Yes, actually, that's exactly the case. Those lines were added in the EOL change noise.

----

As I'm already half done, I'm going to push the rebase to get things sorted out and easier to understand.

----

How exactly are we supposed to use the .clang-format file? Maybe a little explanation about why it exists.

I just played around with the command line tool. It seems to reformat the code according to a rule file. It's caused widespread minor changes to the code base. It seems to have changed some indentation (tabs to spaces maybe?). I noticed it alphabetically sorted include lines. It changes the line wrapping, which didn't always play nice with comments. It tried to justify parameters when wrapping long function signatures. I'm not sure all the changes would be approved by all developers though. The rules might need a little bit of work.

Is there a Visual Studio plugin that uses the format file?

**Comment by cugone** [comment]
@DanRStevens 

Yeah, as mentioned it won't be entirely compatible with the code base. Even the clang-format docs says so:

https://www.kernel.org/doc/html/latest/process/clang-format.html
> clang-format is a tool to format C/C++/... code according to a set of rules and heuristics. Like most tools, **it is not perfect nor covers every single case**, but it is good enough to be helpful.

---

It allows per-project format rules to be applied instead of forcing each developer to change their global format rules.

e.g.
I like spaces and K&R brace style and have set my global format rules to use that. This project uses Allman brace style and tabs. Any time I make a change my global format rule will try to override. Using clang-format allows for per-project format rules to override my global rules and not affect any other unrelated project I'm working on.

---

For VS it is automatically detected and then can optionally be used thereafter. If you choose to use it, it is automatically applied when you complete a Copy-Paste operation or insert a brace/semi-colon. It can also be manually invoked with the shortcut pairs `Ctrl+K, Ctrl+D` (for the entire document) or `Ctrl+K,Ctrl+F` (for the current selection).

I suspect for linux you'll have to take the longer route of manually typing `clang-format [file/directory glob]` into the console.
https://www.kernel.org/doc/html/latest/process/clang-format.html

**Comment by DanRStevens** [comment]
Ok, I just took another look over this thread. I think the changes are ready for merging.

@cugone, your last post nicely summarizes what this file is for, and has links to relevant documentation in case someone needs to update it. It sounds like there is built in Visual Studio support, and hotkeys to activate the functionality.

One point I'm not clear on is if there are visual indicators within Visual Studio about code format violations that this file will enable. Does it provide any kind of real-time syntax-check-like feedback?

----

This thread isn't too long, though we do have a bit of back and forth on the work. As Leeor seems to have limited time and energy for review, perhaps we can help him out a bit with the review process. @cugone, would you be able to open a new PR for this branch? I don't think you need to create a new branch, just a new PR for the same branch. You can use the contents of your last post above as the lead contents for the PR. Feel free to update if you want to add more details on use, or reword things. That should minimize what Leeor has to look over.

Once that's done, please request a review from Leeor (the link is on the upper right of the PR page), and then close the old PR. If you're unable to request a review, please "@" mention him requesting a review.

Thanks for the work.

**Comment by cugone** [comment]
@DanRStevens 
Tried to create a new PR from the same branch and github only has "view the pull request" instead of creating a new one and google isn't helping much.

Unfortunately, I cannot request a review without write access so @ mention it is:
https://help.github.com/en/articles/requesting-a-pull-request-review
 > Note: Pull request authors can't request reviews unless they are either a repository owner or collaborator with write access to the repository.

**Comment by cugone** [comment]
@ldicker83 

Requesting review.

Suggest start from here: https://github.com/lairworks/nas2d-core/pull/128#issuecomment-483044036

**Comment by DanRStevens** [comment]
Maybe the old PR needed to be closed before a new one can be opened. Seems you've found a reasonable compromise though.
